### PR TITLE
Chore Bump python version

### PR DESCRIPTION
**Pull request by tofarr**
We are already using 3.12 in our DockerFile anyway, so this upgrades the python version we use in our makefile

### CODE REVIEW

File: .devcontainer/on_create.sh [hunk 1]
```diff
@@ -2,7 +2,5 @@
 sudo apt update
 sudo apt install -y netcat
 sudo add-apt-repository -y ppa:deadsnakes/ppa
-sudo apt install -y python3.11
-curl -sSL https://install.python-poetry.org | python3.11 -
-# chromadb requires SQLite > 3.35 but SQLite in Python3.11.9 comes with 3.31.1
-sudo cp /opt/conda/lib/libsqlite3.so.0 /lib/x86_64-linux-gnu/libsqlite3.so.0
+sudo apt install -y python3.12
+curl -sSL https://install.python-poetry.org | python3.12 -
```
Comment (tofarr) [inline .devcontainer/on_create.sh:6]: Python 3.12 includes SQLite 3.40

Comment (enyst) [reply .devcontainer/on_create.sh:6]: Interesting. That's a good thing! I hope this works.

FYI, if this PR does work, we probably need to review and/or revert this: https://github.com/All-Hands-AI/OpenHands/pull/3713

**Comment by tobitege** [comment]
@tofarr Hey, you may also have to cherry pick the latest commit from #3791 , my bad :(

**Comment by tobitege** [comment]
I'm investigating in my PR, why the Runtime Units tests work great, but now the integration tests fail 😔

**Comment by enyst** [comment]
For the record, this seems to have been the last time we settled on 3.11: https://github.com/All-Hands-AI/OpenHands/pull/885. It may be worth to test chroma on Macs, as that was the main blocker for upgrading to 3.12 (once this PR is ready for review, we can test fresh installation and upgrade), and we can take a quick look at the linked issues to make sure those issues no longer exist.

FWIW I only use python 3.12 everywhere on openhands, and (1) I experienced the issue with chroma uninstallable via poetry long ago, worked around it manually with pip, and (2) I'm not seeing that issue or any issue related to 3.12 for months, it works just fine.

**Comment by enyst** [comment]
It's still looking for 3.11 🤔 

> openhands:INFO: docker.py:31 - Looking for: ['conda-forge::poetry', 'python=3.11']

**Comment by tobitege** [comment]
Also look into the Dockerfile.j2 template @tofarr
I see like 44 files where 3.11 is mentioned (incl. docs)

File: poetry.lock [hunk 2]
```diff
@@ -5880,32 +5856,21 @@ tests = ["check-manifest", "coverage", "defusedxml", "markdown2", "olefile", "pa
 typing = ["typing-extensions"]
 xmp = ["defusedxml"]
 
-[[package]]
-name = "pip"
-version = "24.2"
```
Comment (enyst) [inline poetry.lock:5885]: This is a bit curious, we had pip and it'll be removed 🤔

File: poetry.lock [hunk 3]
```diff
@@ -441,13 +441,13 @@ dev = ["freezegun (>=1.0,<2.0)", "pytest (>=6.0)", "pytest-cov"]
 
 [[package]]
 name = "backoff"
-version = "2.2.1"
+version = "1.11.1"
```
Comment (enyst) [inline poetry.lock:444]: Not necessarily a problem, I'll note for the record, backoff got downgraded

File: poetry.lock [hunk 4]
```diff
@@ -5404,69 +5398,54 @@ files = [
...
-]
-
-[package.dependencies]
-opentelemetry-proto = "1.26.0"
+importlib-metadata = ">=6.0,<=8.4.0"
 
 [[package]]
 name = "opentelemetry-exporter-otlp-proto-grpc"
-version = "1.26.0"
+version = "1.11.1"
```
Comment (enyst) [inline poetry.lock:5420]: this got downgraded too, and a few more related packages around

Comment (tofarr) [reply poetry.lock:5420]: Interesting - I did a regenerate on this this morning and it bumped it to 1.27.0

File: poetry.lock (hunk 3 above)
Comment (tofarr) [reply poetry.lock:444]: I ran a full `poetry update`, which seems to have resolved this. Not sure why it downgraded...

**Comment by enyst** [comment]
Party time, all CI finally passed! 🥳

**Comment by xingyaoww** [comment]
Are we ready to merge this now all CI passed? 🤔

**Comment by enyst** [comment]
IMHO it would be good to have a few people test it, on WSL / Mac in particular, since those had issues when we last tried this.

It does work on my Mac as upgrade, over the existing stuff. I'll try what a clean install should look like.

Or we can merge and hope for the best, we can also revert if we must. In the first days after merge I assume we'll have reports from folks running `main`.

**Comment by tobitege** [comment]
I could test in WSL/Ubuntu on the weekend

**Comment by tofarr** [comment]
I regenerated this from scratch. It seems to work locally, but the sandbox build was slower due to a lot of dependencies needing different versions.

I also had to constrain protobuf as chromadb / opentelemetry did not work with the latest version.

**Review by rbren — APPROVED** [review]

**Comment by rbren** [comment]
Once the macOS tests pass, ping me and I'll
* force merge this
* change the required checks from 3.11 to 3.12

**Comment by enyst** [comment]
I went through poetry.lock again and didn't see anything suspicious.

Upgraded install works. New install worked (this had an issue on Mac with chroma-hnswlib a few months ago). Curiously though:
```
Collecting chroma-hnswlib
  Using cached chroma_hnswlib-0.7.6-cp312-cp312-macosx_11_0_arm64.whl.metadata (252 bytes)
Collecting numpy (from chroma-hnswlib)
  Downloading numpy-2.1.1-cp312-cp312-macosx_14_0_arm64.whl.metadata (60 kB)
Using cached chroma_hnswlib-0.7.6-cp312-cp312-macosx_11_0_arm64.whl (185 kB)
Downloading numpy-2.1.1-cp312-cp312-macosx_14_0_arm64.whl (5.1 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.1/5.1 MB 20.3 MB/s eta 0:00:00
Installing collected packages: numpy, chroma-hnswlib
Successfully installed chroma-hnswlib-0.7.6 numpy-2.1.1
```
Then:
```
 - Downgrading numpy (2.1.1 -> 1.26.4)
 ```
It looks like `required by ... - llama-index-core <2.0.0` is why. I think it's fine. We have 1.x today and are overdue an update to numpy 2.x.x, but this is not specific to an upgrade to python 3.12.

I ran a `make build / make run` and played around. It looks ok.

**Review by enyst — APPROVED** [review]

**Pull request merged**